File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/EntrySpot/Handlers/play-actor-observe-milestone-handler.ts
import { zlinkEntrySpotActorRequestHandler } from '@zlink-systems/nestjs';
import { observeMilestoneRes, PacketNames } from '../../../../../../../Shared/Contracts/messages';
import type {
  ZLinkEntrySpotActorRequestHandler,
  ZLinkMessageContext
} from '@zlink-systems/framework';
import type {
  ObserveMilestoneReq,
  ObserveMilestoneRes,
} from '../../../../../../../Shared/Contracts/messages';
import { PlayActor } from '../../../Actors/play-actor';
import { MilestoneObserverRegistry } from '../entry-spot-registries';
import { PlayEntrySpot } from '../play-entry-spot';

@zlinkEntrySpotActorRequestHandler({
  actor: () => PlayActor,
  entrySpot: () => PlayEntrySpot,
  packetName: PacketNames.observeMilestoneReq
})
class PlayActorObserveMilestoneHandler
  implements ZLinkEntrySpotActorRequestHandler<
    PlayEntrySpot,
    PlayActor,
    ObserveMilestoneReq,
    ObserveMilestoneRes
  > {
  constructor(private readonly observers: MilestoneObserverRegistry) {}

  async handle(
    _spot: PlayEntrySpot,
    actor: PlayActor,
    context: ZLinkMessageContext,
    _request: ObserveMilestoneReq
  ): Promise<ObserveMilestoneRes> {
    void context;
    this.observers.subscribe(actor.actorId);
    return observeMilestoneRes(true);
  }
}

export { PlayActorObserveMilestoneHandler };

File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/entryspot/handlers/PlayActorObserveMilestoneHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotActorRequest;
import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.PlayEntrySpot;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneReq;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneRes;

@ZLinkHandlerGroup(SampleNames.PlayActor)
public final class PlayActorObserveMilestoneHandler {
    @ZLinkSpotActorRequest
    public CompletionStage<ObserveMilestoneRes> observe(
        PlayEntrySpot entrySpot,
        PlayActor actor,
        ZLinkMessageContext context,
        ObserveMilestoneReq request) {
        return CompletableFuture.completedFuture(entrySpot.observeMilestone(actor));
    }
}
